### PR TITLE
fix state_dict() hanging process with torchmetrics 0.4

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes hanging process due to state_dict() not being called on all ranks. 
This is necessary after torchmetrics 0.4, PR https://github.com/PyTorchLightning/metrics/pull/302 
PL master is fine.
This fix will be included in the weekly patch release #8198 

Additional context: https://github.com/PyTorchLightning/metrics/issues/331

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 1]
```diff
@@ -196,7 +199,7 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
                 and len(best_model_path) > 0
             ):
                 last_path = re.sub(".ckpt", ".tmp_end.ckpt", best_model_path)
-                self.save(self.lightning_module.state_dict(), last_path)
+                self.save(state_dict, last_path)
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:202]: Changes were not required for `tpu_spawn`. As `self.save` is called for all ranks, but this looks good and clean as well.

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:202]: Okay, thanks. Then you recommend to still keep it like this?

Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:202]: Yup, the change is fine.

**Comment by awaelchli** [comment]
@Borda I think we need to combine our two PRs #8206
Mine fixes the hanging process, your's fixes the api compatibility. How do we want to proceed?

**Comment by Borda** [comment]
> Mine fixes the hanging process, your's fixes the api compatibility. How do we want to proceed?

I think we shall merge both kinds of independently (do not matter on order) anyway we need Admin to do this merge... :]

**Comment by kaushikb11** [comment]
@Borda Admins are in a different timezone!

**Comment by awaelchli** [comment]
We asked admin to merge this one with failing tests, then #8206 will follow.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
